Remove mbedtls_rsa_init() padding parameters

ronald-cron-arm opened the pull request:
## Description
Fix #4523

## Status
Ready

## Requires Backporting
NO, 3.0 only.  

## Todos
- [x] Tests
- [x] Documentation
- [x] Changelog updated
- [x] Migration guide

gilles-peskine-arm commented on include/mbedtls/rsa.h:
```diff
@@ -134,33 +134,51 @@ mbedtls_rsa_context;
- * \param hash_id  The hash identifier of ::mbedtls_md_type_t type, if
- *                 \p padding is #MBEDTLS_RSA_PKCS_V21. It is unused
- *                 otherwise.
+ * \param hash_id  The #MBEDTLS_RSA_PKCS_V21 hash identifier.
```
This sentence is misleading: it looks like the value here should be `MBEDTLS_RSA_PKCS_V21`. Maybe this?

> The hash identifier for PSS or OAEP, if `\p padding` is `#MBEDTLS_RSA_PKCS_V21`. `#MBEDTLS_MD_NONE` …. Ignored if `\p padding` is `#MBEDTLS_RSA_PKCSV15`.

gilles-peskine-arm commented on programs/pkey/rsa_sign_pss.c:
```diff
@@ -115,7 +115,13 @@ int main( int argc, char *argv[] )
+                                         MBEDTLS_RSA_PKCS_V21,
+                                         MBEDTLS_MD_SHA256 ) ) != 0 )
+    {
+        mbedtls_printf( " failed\n  ! Invalid padding\n" );
```
Minor: we know that our code is correct so this can't fail. “Padding not supported” would be a better message, although at this time `mbedtls_rsa_set_padding` doesn't actually fail in this case.

gilles-peskine-arm commented on tests/suites/test_suite_pkcs1_v15.function:
```diff
@@ -23,7 +23,8 @@ void pkcs1_rsaes_v15_encrypt( int mod, int radix_N, char * input_N,
     mbedtls_mpi_init( &N ); mbedtls_mpi_init( &E );
-    mbedtls_rsa_init( &ctx, MBEDTLS_RSA_PKCS_V15, hash );
+    mbedtls_rsa_init( &ctx );
+    mbedtls_rsa_set_padding( &ctx, MBEDTLS_RSA_PKCS_V15, hash );
```
In the unit tests, please change all calls to `mbedtls_rsa_set_padding` to assert success.

gilles-peskine-arm commented on library/rsa.c:
```diff
@@ -500,15 +497,27 @@ void mbedtls_rsa_init( mbedtls_rsa_context *ctx,
-    RSA_VALIDATE( ctx != NULL );
-    RSA_VALIDATE( padding == MBEDTLS_RSA_PKCS_V15 ||
-                  padding == MBEDTLS_RSA_PKCS_V21 );
+    if( ( padding != MBEDTLS_RSA_PKCS_V15 ) &&
+        ( padding != MBEDTLS_RSA_PKCS_V21 ) )
+        return( MBEDTLS_ERR_RSA_INVALID_PADDING );
```
I think it would be better to fail immediately if the chosen padding type is disabled by the build-time configuration.

ronald-cron-arm commented on ChangeLog.d/rsa-padding.txt:
```diff
@@ -0,0 +1,7 @@
+API changes
+   * It is not possible anymore to select the PKCS#1 v2.1 encoding and its
+     hash with mbedtls_rsa_init() which now selects by default the PKCS#1
+     v1.5 encoding. The selection of PKCS#1 v2.1 encoding and its hash can now
+     only be made with mbedtls_rsa_set_padding(). In addition,
+     mbedtls_rsa_set_padding() now returns an error code if the padding
+     parameters are invalid.
```
Yes better, changed.

ronald-cron-arm commented:
@gilles-peskine-arm I believe I have addressed all comments. There is a conflict now with `development` due to the merge of #4516 I believe. Let me know how as a reviewer you prefer me to handle that.

gilles-peskine-arm commented:
Please go ahead and rebase.

ronald-cron-arm commented:
I've rebased onto `development`. The version previous to the rebase is [here](https://github.com/ronald-cron-arm/mbedtls/tree/rsa-padding-1). The conflicts were all about the rsa_invalid_param( ) test function that was removed from `development` by #4516 and that this PR reintroduces.

gilles-peskine-arm reviewed (dismissed):
LGTM except for a hard-to-read bit in the migration guide.

davidhorstmann-arm reviewed (dismissed):
LGTM

ronald-cron-arm commented on docs/3.0-migration-guide.d/rsa-padding.md:
```diff
@@ -0,0 +1,29 @@
+you were using the PKCS#1 v2.1 encoding you now need, subsequently to the call
+to mbedtls_rsa_init(), to call mbedtls_rsa_set_padding() to set it.
+
+Code migration examples:
```
I've tried to improve it following your suggestion. The layout with the blocks of code is still a bit weird but for sure better.

gilles-peskine-arm commented on docs/3.0-migration-guide.d/rsa-padding.md:
````diff
@@ -0,0 +1,31 @@
+```C
+    mbedtls_rsa_init(ctx);
+```
+.
````
There shouldn't be a lone punctuation after a code block (or displayed equation).

ronald-cron-arm replied:
Yes I am not comfortable with either solution anyway. Removing.

ronald-cron-arm replied:
I've amended the last commit.

davidhorstmann-arm reviewed (approved):
LGTM
